### PR TITLE
vendor: use go mod (rebase after removal of deprecated host-discovery)

### PR DESCRIPTION
**Pull request by thaJeztah**
this is a rebase of https://github.com/moby/moby/pull/43075 to see if it's easy to re-do the go mod changes if https://github.com/moby/moby/pull/42247 would be merged first

I made some changes in separate commits (could be squashed); using this draft PR to check CI.

### CODE REVIEW

**Comment by thaJeztah** [comment]
Hmm.. interesting; looks like vendor-check fails (diff in `vendor.sum`), but doesn't stop CI from running (so it keeps hanging on something);

```
[2021-12-22T17:40:44.561Z] patching file stat_unix.go
[2021-12-22T17:40:44.561Z] Hunk #1 succeeded at 9 (offset 1 line).
[2021-12-22T17:40:44.561Z] Hunk #2 succeeded at 17 (offset 1 line).
[2021-12-22T17:40:44.561Z] Hunk #3 succeeded at 25 (offset 1 line).
[2021-12-22T17:40:44.561Z] The result of go mod vendor differs
[2021-12-22T17:40:44.561Z] 
[2021-12-22T17:40:44.561Z]  M vendor.sum
[2021-12-22T17:40:44.561Z] 
[2021-12-22T17:40:44.561Z] Please vendor your package with hack/vendor.sh.
[2021-12-22T17:40:44.561Z] 
[2021-12-22T17:40:44.561Z] diff --git a/vendor.sum b/vendor.sum
[2021-12-22T17:40:44.561Z] index 2b8477107a..dc3747792d 100644
[2021-12-22T17:40:44.561Z] --- a/vendor.sum
[2021-12-22T17:40:44.561Z] +++ b/vendor.sum
[2021-12-22T17:40:44.561Z] @@ -123,6 +123,7 @@ github.com/coreos/etcd v3.3.25+incompatible/go.mod h
[2021-12-22T17:40:44.561Z] 1:uF7uidLiAD3TWHmW31ZFd/JWoc
[2021-12-22T17:40:44.561Z]  github.com/coreos/go-iptables v0.5.0/go.mod h1:/mVI274lEDI2ns62jHCDnCyBF9Iwsmek
[2021-12-22T17:40:44.561Z] av8Dbxlm1MU=
[2021-12-22T17:40:44.561Z]  github.com/coreos/go-oidc v2.1.0+incompatible/go.mod h1:CgnwVTmzoESiwO9qyAFEMiH
[2021-12-22T17:40:44.561Z] oZ1nMCKZlZ9V6mm3/LKc=
[2021-12-22T17:40:44.561Z]  github.com/coreos/go-semver v0.2.0/go.mod h1:nnelYz7RCh+5ahJtPPxZlU+153eP4D4r3E
[2021-12-22T17:40:44.561Z] edlOD2RNk=
[2021-12-22T17:40:44.561Z] +github.com/coreos/go-semver v0.3.0/go.mod h1:nnelYz7RCh+5ahJtPPxZlU+1
[2021-12-22T17:40:44.561Z] 53eP4D4r3EedlOD2RNk=
[2021-12-22T17:40:44.561Z]  github.com/coreos/go-systemd v0.0.0-20180511133405-39ca1b05acc7 h1:u9SHYsPQNyt5
[2021-12-22T17:40:44.561Z] tgDm3YN7+9dYrpK96E5wFilTFWIDZOM=
[2021-12-22T17:40:44.561Z]  github.com/coreos/go-systemd v0.0.0-20180511133405-39ca1b05acc7/go.mod h1:F5haX
[2021-12-22T17:40:44.561Z] 7vjVVG0kc13fIWeqUViNPyEJxv/OmvnBo0Yme4=
[2021-12-22T17:40:44.561Z]  github.com/coreos/go-systemd/v22 v22.1.0/go.mod h1:xO0FLkIi5MaZafQlIrOotqXZ90ih
[2021-12-22T17:40:44.561Z] +1atmu1JpKERPPk=
[2021-12-22T17:40:44.561Z] @@ -205,7 +206,6 @@ github.com/go-openapi/spec v0.19.3/go.mod h1:FpwSN1k
[2021-12-22T17:40:44.561Z] sY1eteniUU7X0N/BgJ7a4WvBFVA8
[2021-12-22T17:40:44.561Z]  github.com/go-openapi/swag v0.19.2/go.mod h1:POnQmlKehdgb5mhVOsnJFsivZCEZ/vjK9g
```

**Comment by thaJeztah** [comment]
Probably because `hack/vendor.sh` doesn't run `go mod tidy` (and I may have forgotten to run it again 🤔)

**Comment by thaJeztah** [comment]
Hmm.. new failure?

```
=== RUN   TestFollowLogsHandleDecodeErr
time="2021-12-23T11:18:44Z" level=error msg="Log reader notified that it must re-open log file, some log data may not be streamed to the client." file=/tmp/TestFollowLogsHandleDecodeErr3941852975
    logfile_test.go:317: assertion failed: 5 (int) != 6 (dec.resetCount int)
--- FAIL: TestFollowLogsHandleDecodeErr (0.01s)
```

File: vendor.mod [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	github.com/BurntSushi/toml v0.3.1 // indirect
+	github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da // indirect
+	github.com/beorn7/perks v1.0.1 // indirect
+	github.com/bmizerany/assert v0.0.0-20160611221934-b7ed37b82869 // indirect
+	github.com/cespare/xxhash/v2 v2.1.1 // indirect
+	github.com/cilium/ebpf v0.6.2 // indirect
+	github.com/containerd/console v1.0.2 // indirect
+	github.com/containerd/go-runc v1.0.0 // indirect
+	github.com/containerd/ttrpc v1.1.0 // indirect
+	github.com/coreos/etcd v3.3.25+incompatible // indirect
```
Comment (thaJeztah) [inline vendor.mod:102]: Note to self; I need to update this to the version that's used in vendor.conf (see https://github.com/moby/moby/pull/41791)

File: vendor.mod [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	github.com/docker/distribution v2.7.1+incompatible
+	github.com/docker/docker/autogen/winresources/dockerd v0.0.0-00010101000000-000000000000
+	github.com/docker/go-connections v0.4.0
+	github.com/docker/go-events v0.0.0-20190806004212-e31b211e4f1c
+	github.com/docker/go-metrics v0.0.1
+	github.com/docker/go-units v0.4.0
+	github.com/docker/libkv v0.2.2-0.20211217103745-e480589147e3
+	github.com/docker/libtrust v0.0.0-20150526203908-9cbd2a1374f4
+	github.com/docker/swarmkit v1.12.1-0.20210726173615-3629f50980f6
+	github.com/fluent/fluent-logger-golang v1.8.0
```
Comment (thaJeztah) [inline vendor.mod:37]: Note to self: need to update this to v1.9.0 (see https://github.com/moby/moby/pull/43100)

**Review by errordeveloper — APPROVED** [review]

**Comment by errordeveloper** [comment]
@thaJeztah thanks for rebasing and the additions!

**Comment by thaJeztah** [comment]
Weird failures on Windows 😞 (likely unrelated)

```
SUCCESS: Specified value was saved.
INFO: Downloading git...
INFO: Downloading go...
INFO: Downloading compiler 1 of 3...
INFO: Downloading compiler 2 of 3...
INFO: Downloading compiler 3 of 3...
INFO: Extracting git...
Move-Item : Access to the path 'C:\git-tmp\tools\usr' is denied.
At line:1 char:2427
+ ... tory C:\git | Out-Null; Move-Item C:\git-tmp\tools\* C:\git\.; Remove ...
+                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (C:\git-tmp\tools\usr:DirectoryInfo) 
    [Move-Item], IOException
    + FullyQualifiedErrorId : MoveDirectoryItemIOError,Microsoft.PowerShell.Commands.MoveItemCommand
```

**Comment by errordeveloper** [comment]
@thaJeztah I have seen similar Windows file access/modification races on my PR earlier... definitely unrelated.

**Comment by thaJeztah** [comment]
Looking mostly OK; Other failure on Windows 2022 is a known flaky; https://github.com/moby/moby/issues/43012

```
=== FAIL: github.com/docker/docker/integration-cli TestDockerSuite/TestSlowStdinClosing (1.08s)
    docker_cli_run_test.go:4177: assertion failed: error is not nil: exit status 127
    --- FAIL: TestDockerSuite/TestSlowStdinClosing (1.08s)
```

**Comment by errordeveloper** [comment]
@thaJeztah should this be marked "ready for review"?

**Review by AkihiroSuda — APPROVED** [review]

**Review by tonistiigi — APPROVED** [review]

**Comment by thaJeztah** [comment]
All green now! We rolled back the windows machines to the previous builds (gonna be using https://github.com/moby/moby/pull/43145 to track if we're able to find where the regression is in Windows; looks like there's issues with newer builds of Windows itself that cause CI to fail and to run severely slower, but we may need help from the Microsoft people on that, as it looks to be in the platform itself)

**Comment by thaJeztah** [comment]
Next steps will be to remove the `replace` rules, and to update dependencies (containerd, buildkit, swarmkit)

**Comment by slonopotamus** [comment]
So, is moby moving towards using Go modules in a proper way, with go.mod/go.sum?

**Comment by thaJeztah** [comment]
Ultimately, yes, but that requires quite some additional changes to make it work

**Pull request merged**